### PR TITLE
Add bridge from complex in Zeros to real/imag in Zeros

### PR DESCRIPTION
**Pull request by blegat**
Refactored from https://github.com/jump-dev/ComplexOptInterface.jl

### CODE REVIEW

File: src/Bridges/Constraint/Constraint.jl [hunk 1]
```diff
@@ -7,6 +7,8 @@
 module Constraint
 
 import LinearAlgebra
+import MutableArithmetics
+const MA = MutableArithmetics
```
Comment (odow) [inline src/Bridges/Constraint/Constraint.jl:11]: I'm going to enforce my rule of no new acronyms other than MOI. They just make the code harder to read. You should be able to read each bridge file in isolation without needing to go looking for what each thing is. I know that _we_ know what MA is, but it just makes things harder for others coming along later.

File: src/Bridges/Constraint/bridges/split_zeros.jl [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+const SplitZeros{T,OT<:MOI.ModelLike} =
+    SingleBridgeOptimizer{SplitZerosBridge{T},OT}
+
+function _nonzero_indices(func::MOI.AbstractVectorFunction)
+    return [
+        i for (i, scalar_func) in enumerate(MOI.Utilities.scalarize(func)) if
+        !iszero(scalar_func)
+    ]
+end
+function bridge_constraint(
```
Comment (odow) [inline src/Bridges/Constraint/bridges/split_zeros.jl:48]:
```suggestion

function bridge_constraint(
```

File: src/Bridges/Constraint/bridges/split_zeros.jl [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+# We don't support `MOI.VectorOfVariables` as it would be a self-loop in the bridge graph
+function MOI.supports_constraint(
+    ::Type{SplitZerosBridge{T}},
+    ::Type{<:MOI.Utilities.TypedLike{Complex{T}}},
+    ::Type{MOI.Zeros},
+) where {T}
+    return true
+end
+function MOI.Bridges.added_constrained_variable_types(
```
Comment (odow) [inline src/Bridges/Constraint/bridges/split_zeros.jl:85]:
```suggestion

function MOI.Bridges.added_constrained_variable_types(
```

File: src/Bridges/Constraint/bridges/split_zeros.jl [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    ::Type{<:SplitZerosBridge},
+)
+    return Tuple{DataType}[]
+end
+function MOI.Bridges.added_constraint_types(
```
Comment (odow) [inline src/Bridges/Constraint/bridges/split_zeros.jl:90]:
```suggestion

function MOI.Bridges.added_constraint_types(
```

File: src/Bridges/Constraint/bridges/split_zeros.jl [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    ::Type{SplitZerosBridge{T,F,G}},
+) where {T,F,G}
+    return Tuple{DataType,DataType}[(F, MOI.Zeros)]
+end
+function concrete_bridge_type(
```
Comment (odow) [inline src/Bridges/Constraint/bridges/split_zeros.jl:95]:
```suggestion

function concrete_bridge_type(
```

File: src/Bridges/Constraint/bridges/split_zeros.jl [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    ::Type{<:SplitZerosBridge{T}},
+    G::Type{<:MOI.Utilities.TypedLike},
+    ::Type{MOI.Zeros},
+) where {T}
+    F = MA.promote_operation(imag, G)
```
Comment (odow) [inline src/Bridges/Constraint/bridges/split_zeros.jl:100]:
```suggestion
    F = MutableArithmetics.promote_operation(imag, G)
```

File: src/Bridges/Constraint/bridges/split_zeros.jl [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+end
+
+# Attributes, Bridge acting as a model
+function MOI.get(
+    ::SplitZerosBridge{T,F},
+    ::MOI.NumberOfConstraints{F,MOI.Zeros},
+) where {T,F}
+    return 1
+end
+function MOI.get(
```
Comment (odow) [inline src/Bridges/Constraint/bridges/split_zeros.jl:111]:
```suggestion

function MOI.get(
```

File: src/Bridges/Constraint/bridges/split_zeros.jl [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+) where {T}
+    F = MA.promote_operation(imag, G)
+    return SplitZerosBridge{T,F,G}
+end
+
+# Attributes, Bridge acting as a model
+function MOI.get(
+    ::SplitZerosBridge{T,F},
+    ::MOI.NumberOfConstraints{F,MOI.Zeros},
+) where {T,F}
```
Comment (odow) [inline src/Bridges/Constraint/bridges/split_zeros.jl:108]:
```suggestion
)::Int64 where {T,F}
```

File: src/Bridges/Constraint/bridges/split_zeros.jl [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+function concrete_bridge_type(
+    ::Type{<:SplitZerosBridge{T}},
+    G::Type{<:MOI.Utilities.TypedLike},
+    ::Type{MOI.Zeros},
+) where {T}
+    F = MA.promote_operation(imag, G)
+    return SplitZerosBridge{T,F,G}
+end
+
+# Attributes, Bridge acting as a model
```
Comment (odow) [inline src/Bridges/Constraint/bridges/split_zeros.jl:104]:
```suggestion
```

File: src/Bridges/Constraint/bridges/split_zeros.jl [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+function MOI.get(
+    bridge::SplitZerosBridge{T,F},
+    ::MOI.ListOfConstraintIndices{F,MOI.Zeros},
+) where {T,F}
+    return [bridge.constraint]
+end
+
+# Indices
+function MOI.delete(model::MOI.ModelLike, bridge::SplitZerosBridge)
+    return MOI.delete(model, bridge.constraint)
```
Comment (odow) [inline src/Bridges/Constraint/bridges/split_zeros.jl:120]:
```suggestion
    MOI.delete(model, bridge.constraint)
    return
```

File: src/Bridges/Constraint/bridges/split_zeros.jl [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+    return 1
+end
+function MOI.get(
+    bridge::SplitZerosBridge{T,F},
+    ::MOI.ListOfConstraintIndices{F,MOI.Zeros},
+) where {T,F}
+    return [bridge.constraint]
+end
+
+# Indices
```
Comment (odow) [inline src/Bridges/Constraint/bridges/split_zeros.jl:118]:
```suggestion
```

File: src/Bridges/Constraint/bridges/split_zeros.jl [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+function MOI.delete(model::MOI.ModelLike, bridge::SplitZerosBridge)
+    return MOI.delete(model, bridge.constraint)
+end
+
+# Attributes, Bridge acting as a constraint
```
Comment (odow) [inline src/Bridges/Constraint/bridges/split_zeros.jl:123]:
```suggestion
```

File: src/Bridges/Constraint/bridges/split_zeros.jl [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+function MOI.supports(
+    ::MOI.ModelLike,
+    ::Union{MOI.ConstraintPrimalStart,MOI.ConstraintDualStart},
+    ::Type{<:SplitZerosBridge},
+)
+    return true
+end
+function MOI.get(
```
Comment (odow) [inline src/Bridges/Constraint/bridges/split_zeros.jl:131]:
```suggestion

function MOI.get(
```

File: src/Bridges/Constraint/bridges/split_zeros.jl [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+    output = zeros(Complex{eltype(values)}, bridge.dimension)
+    for (i, idx) in enumerate(bridge.real_indices)
+        output[idx] = values[i]
+    end
+    for (i, idx) in enumerate(bridge.imag_indices)
+        output[idx] = values[length(bridge.real_indices)+i] * im
+    end
+    return output
+end
+function MOI.set(
```
Comment (odow) [inline src/Bridges/Constraint/bridges/split_zeros.jl:151]:
```suggestion

function MOI.set(
```

File: test/Bridges/Constraint/split_zeros.jl [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+            (MOI.VectorAffineFunction{T}, MOI.Zeros) => [zeros(T, 1)],
+        ),
+    )
+    MOI.Test.test_linear_complex_Zeros_duplicate(bridged_mock, config)
+    cis = MOI.get(
+        mock,
+        MOI.ListOfConstraintIndices{MOI.VectorAffineFunction{T},MOI.Zeros}(),
+    )
+    @test length(cis) == 1
+end
```
Comment (odow) [inline test/Bridges/Constraint/split_zeros.jl:71]:
```suggestion
    return
end
```

File: src/Bridges/Constraint/Constraint.jl (hunk 1 above)
Comment (odow) [inline src/Bridges/Constraint/Constraint.jl:11]:
```suggestion
```

**Review by odow — COMMENTED** [review]
Does a `MOI.Bridges.runtest` work with complex numbers?

**Comment by blegat** [comment]
> Does a `MOI.Bridges.runtest` work with complex numbers?

The parser only works with `Float64` input, see e.g.
https://github.com/jump-dev/MathOptInterface.jl/blob/868aa29bffaa6fd727a1858897ef961cc7b34e12/src/Utilities/parser.jl#L10
and since the expression `1.0im` is a product and not a number, the following fails as well:
https://github.com/jump-dev/MathOptInterface.jl/blob/868aa29bffaa6fd727a1858897ef961cc7b34e12/src/Utilities/parser.jl#L139

File: src/Bridges/Constraint/bridges/split_zeros.jl [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+`SplitZerosBridge` implements the following reformulation:
+
+  * ``f(x) \\in \\{0\\}^n`` into ``\\text{Re}(f(x)) \\in \\{0\\}^n`` and ``\\text{Im}(f(x)) \\in \\{0\\}^n``
+
+## Source node
+
+`SplitZerosBridge` supports:
+
+  * `G` in [`MOI.Zeros`](@ref)
+
```
Comment (odow) [inline src/Bridges/Constraint/bridges/split_zeros.jl:19]:
```suggestion

where `G` is a function with `Complex` coefficients.

```

File: src/Bridges/Constraint/bridges/split_zeros.jl [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+        MOI.dimension(set),
+        constraint,
+        real_indices,
+        imag_indices,
+    )
+end
+
+# We don't support `MOI.VectorOfVariables` as it would be a self-loop in the bridge graph
+function MOI.supports_constraint(
+    ::Type{SplitZerosBridge{T}},
```
Comment (odow) [inline src/Bridges/Constraint/bridges/split_zeros.jl:80]:
```suggestion
    ::Type{<:SplitZerosBridge{T}},
```

**Review by odow — APPROVED** [review]
Okay for now. I might take a go at adding complex support to the model parser.

**Pull request merged**